cloudinit/sources/DataSourceVMware.py

```diff
@@ -0,0 +1,814 @@
+
+    try:
+        subp(
+            [
+                vmware_rpctool,
+                ("info-set %s %s" % (get_guestinfo_key_name(key), value)),
+            ]
+        )
+        return True
+    except ProcessExecutionError as error:
+        util.logexc(
+            LOG,
+            "Failed to set guestinfo key=%s to value=%s: %s",
+            key,
+            value,
+            error,
+        )
+    except Exception:
+        util.logexc(
+            LOG,
+            "Unexpected error while trying to set "
+            + "guestinfo key=%s to value=%s",
+            key,
+            value,
+        )
+
+    return None
+
+
+def guestinfo_redact_keys(keys, vmware_rpctool=VMWARE_RPCTOOL):
+    """
+    guestinfo_redact_keys redacts guestinfo of all of the keys in the given
+    list. each key will have its value set to "---". Since the value is valid
+    YAML, cloud-init can still read it if it tries.
+    """
+    if not keys:
+        return
+    if not type(keys) in (list, tuple):
+        keys = [keys]
+    for key in keys:
+        key_name = get_guestinfo_key_name(key)
+        LOG.info("clearing %s", key_name)
+        if not guestinfo_set_value(
+            key, GUESTINFO_EMPTY_YAML_VAL, vmware_rpctool
+        ):
+            LOG.error("failed to clear %s", key_name)
+        LOG.info("clearing %s.encoding", key_name)
+        if not guestinfo_set_value(key + ".encoding", "", vmware_rpctool):
+            LOG.error("failed to clear %s.encoding", key_name)
+
+
+def load(data):
+    """
+    load first attempts to unmarshal the provided data as JSON, and if
+    that fails then attempts to unmarshal the data as YAML. If data is
+    None then a new dictionary is returned.
+    """
+    if not data:
+        return {}
+    try:
+        return util.load_json(data)
+    except (json.JSONDecodeError, TypeError):
+        return util.load_yaml(data)
+
+
+def process_metadata(data):
+    """
+    process_metadata processes metadata and loads the optional network
+    configuration.
+    """
+    network = None
+    if "network" in data:
+        network = data["network"]
+        del data["network"]
+
+    network_enc = None
+    if "network.encoding" in data:
+        network_enc = data["network.encoding"]
+        del data["network.encoding"]
+
+    if network:
+        if isinstance(network, collections.abc.Mapping):
+            LOG.debug("network data copied to 'config' key")
+            network = {"config": copy.deepcopy(network)}
+        else:
+            LOG.debug("network data to be decoded %s", network)
+            dec_net = decode("metadata.network", network_enc, network)
+            network = {
+                "config": load(dec_net),
+            }
+
+        LOG.debug("network data %s", network)
+        data["network"] = network
+
+    return data
+
+
+def get_datasource_list(depends):
+    """
+    Return a list of data sources that match this set of dependencies
+    """
+    return [DataSourceVMware]
```

Comment:
To avoid accidentally injecting the VMware datasource into a not-yet-defined cloud-init boot stage, can we adhere to the common practice of defining the mapping of authorized boot stages where DataSourceVmware can be discovered?
This way we don't find out later if we add a stage before init-local or init that we are spending time trying to detect VMware where it can't possibly run.
```suggestion
# Used to match classes to dependencies                                         
datasources = [                                                                 
    (DataSourceVMware, (sources.DEP_FILESYSTEM,)),  # Run at init-local       
    (DataSourceVMware, (sources.DEP_FILESYSTEM, sources.DEP_NETWORK)),             
]                                                                               
                                                                                
                                                                                
def get_datasource_list(depends):                                               
    """
    Return a list of data sources that match this set of dependencies
    """
    return sources.list_from_depends(depends, datasources)   
```

Reply:
So I wondered about that. I took the approach that since the DS didn't need the network to get its metadata, that there was no difference to the DS as to why it was loaded.

The examples I saw such as EC2 used the FS level in order to bring up the ephemeral DHCP net to retrieve the metadata on the 169 address.